File: odbc_tests/tests/e2e/authentication/oauth.cpp
#include <picojson.h>
#include <sql.h>
#include <sqlext.h>
#include <sqltypes.h>

#include <sstream>
#include <string>

#include <catch2/catch_test_macros.hpp>

#include "HandleWrapper.hpp"
#include "compatibility.hpp"
#include "get_diag_rec.hpp"
#include "oauth_auth_helpers.hpp"
#include "require.hpp"
#include "test_setup.hpp"

// End-to-end OAuth tests for the ODBC wrapper, mirroring the Python suite
// (python/tests/e2e/authentication/test_oauth.py and auth_helpers.py) and
// sf_core's `oauth_should_*` methods against the shared Gherkin scenarios in
// tests/definitions/shared/authentication/oauth.feature.
//
// Every test needs the headless-browser Docker container
// (snowdrivers-test-external-browser-universal-driver) -- for the Authorization
// Code browser leg and for preprod Okta IdP connectivity -- so all are gated
// behind REQUIRE_BROWSER. Rather than reading a pre-acquired token, the legacy
// OAUTH flow mints a fresh access token from Okta and the Authorization Code
// flow drives the container's provideBrowserCredentials.js. The required
// parameters.json keys are listed inline at each test's get_param_required call.

namespace {

EnvironmentHandleWrapper setup_oauth_environment() {
  // unixODBC reads ODBCSYSINI when the environment handle is allocated, so the driver
  // must be registered before SQLAllocHandle(SQL_HANDLE_ENV).
  ensure_driver_installed();
  EnvironmentHandleWrapper env;
  SQLRETURN ret = SQLSetEnvAttr(env.getHandle(), SQL_ATTR_ODBC_VERSION, (SQLPOINTER)SQL_OV_ODBC3, 0);
  REQUIRE_ODBC(ret, env);
  return env;
}

std::string get_oauth_base_connection_string(const picojson::object& params, const std::string& authenticator,
                                             const std::string& uid) {
  std::stringstream ss;
  configure_driver_string(ss);
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_HOST", "SERVER");
  add_param_required<std::string>(ss, params, "SNOWFLAKE_TEST_ACCOUNT", "ACCOUNT");
  ss << "UID=" << uid << ";";
  add_param_optional<std::string>(ss, params, "SNOWFLAKE_TEST_WAREHOUSE", "WAREHOUSE");
  // ROLE required, else login fails with "No default role assigned" (390194).
  add_param_optional<std::string>(ss, params, "SNOWFLAKE_TEST_ROLE", "ROLE");
  add_param_optional<std::string>(ss, params, "SNOWFLAKE_TEST_DATABASE", "DATABASE");
  add_param_optional<std::string>(ss, params, "SNOWFLAKE_TEST_SCHEMA", "SCHEMA");
  ss << "AUTHENTICATOR=" << authenticator << ";";
  return ss.str();
}

void verify_oauth_simple_query_execution(ConnectionHandleWrapper& dbc) {
  StatementHandleWrapper stmt = dbc.createStatementHandle();
  SQLRETURN ret = SQLExecDirect(stmt.getHandle(), (SQLCHAR*)"SELECT 1", SQL_NTS);
  REQUIRE_ODBC(ret, stmt);

  ret = SQLFetch(stmt.getHandle());
  REQUIRE_ODBC(ret, stmt);

  SQLINTEGER result = 0;
  ret = SQLGetData(stmt.getHandle(), 1, SQL_C_LONG, &result, sizeof(result), nullptr);
  REQUIRE_ODBC(ret, stmt);
  REQUIRE(result == 1);
}

}  // anonymous namespace

// =============================================================================
// Legacy AUTHENTICATOR=OAUTH (pre-acquired access token)
//
// A fresh OAuth access token is minted from the Okta IdP and passed via
// TOKEN; it is presented to Snowflake as-is.
// =============================================================================

TEST_CASE("oauth should authenticate with pre acquired access token", "[oauth_e2e][requires_browser]") {
  REQUIRE_BROWSER("OAuth E2E needs the headless browser container (preprod IdP connectivity)");

  auto params = get_test_parameters("testconnection");
  const std::string user = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_USER");
  const std::string access_token = oauth_auth::retrieve_oauth_access_token(
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_TOKEN_URL"),
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_CLIENT_ID"),
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_CLIENT_SECRET"), user,
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_PASSWORD"),
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_ROLE"));

  // Given Authentication is set to legacy OAUTH and a pre-acquired
  //       OAuth access token is supplied via `token=`
  std::stringstream ss;
  ss << get_oauth_base_connection_string(params, "OAUTH", user);
  ss << "TOKEN=" << access_token << ";";
  std::string connection_string = ss.str();

  auto env = setup_oauth_environment();
  auto dbc = env.createConnectionHandle();

  // When Trying to Connect
  SQLRETURN ret = SQLDriverConnect(dbc.getHandle(), nullptr, (SQLCHAR*)connection_string.c_str(), SQL_NTS, nullptr, 0,
                                   nullptr, SQL_DRIVER_NOPROMPT);
  REQUIRE_ODBC(ret, dbc);

  // Then Login is successful and a simple query can be executed
  verify_oauth_simple_query_execution(dbc);

  SQLRETURN disconnect_ret = SQLDisconnect(dbc.getHandle());
  REQUIRE_ODBC(disconnect_ret, dbc);
}

TEST_CASE("oauth should authenticate using lowercase oauth authenticator", "[oauth_e2e][requires_browser]") {
  REQUIRE_BROWSER("OAuth E2E needs the headless browser container (preprod IdP connectivity)");

  auto params = get_test_parameters("testconnection");
  const std::string user = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_USER");
  const std::string access_token = oauth_auth::retrieve_oauth_access_token(
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_TOKEN_URL"),
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_CLIENT_ID"),
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_CLIENT_SECRET"), user,
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_PASSWORD"),
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_ROLE"));

  // Given Authentication is set to lowercase oauth and a valid pre-acquired OAuth access token is supplied via TOKEN
  std::stringstream ss;
  ss << get_oauth_base_connection_string(params, "oauth", user);
  ss << "TOKEN=" << access_token << ";";
  std::string connection_string = ss.str();

  auto env = setup_oauth_environment();
  auto dbc = env.createConnectionHandle();

  // When Trying to Connect
  SQLRETURN ret = SQLDriverConnect(dbc.getHandle(), nullptr, (SQLCHAR*)connection_string.c_str(), SQL_NTS, nullptr, 0,
                                   nullptr, SQL_DRIVER_NOPROMPT);
  REQUIRE_ODBC(ret, dbc);

  // Then Login is successful and a simple query can be executed
  verify_oauth_simple_query_execution(dbc);

  SQLRETURN disconnect_ret = SQLDisconnect(dbc.getHandle());
  REQUIRE_ODBC(disconnect_ret, dbc);
}

TEST_CASE("oauth should fail legacy authentication with invalid token", "[oauth_e2e][requires_browser]") {
  SKIP("Disabled: bad-secret tests cause pipeline flakiness by blocking the test account");
  REQUIRE_BROWSER("OAuth E2E needs the headless browser container (preprod IdP connectivity)");

  auto params = get_test_parameters("testconnection");
  const std::string user = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_USER");

  // Given Authentication is set to legacy OAUTH and an invalid
  //       OAuth access token is supplied
  std::stringstream ss;
  ss << get_oauth_base_connection_string(params, "OAUTH", user);
  ss << "TOKEN=invalid_oauth_token_12345;";
  std::string connection_string = ss.str();

  ensure_driver_installed();

  // When Trying to Connect
  auto records = require_connection_failed(connection_string);

  // Then Connection fails with an authentication / login error
  REQUIRE(records.size() >= 1);
  CHECK(records[0].sqlState == "28000");
}

// =============================================================================
// OAuth Authorization Code (AC) flow
//
// An interactive, user-based flow that authenticates a real user through a
// browser login leg. The connect thread spawns Chromium via the OS browser
// opener; the browser thread drives the Snowflake IdP login over Chromium's
// remote-debugging port.
// =============================================================================

TEST_CASE("oauth should authenticate using authorization code flow", "[oauth_e2e][requires_browser]") {
  REQUIRE_BROWSER("Authorization Code happy path needs the headless Chromium container");

  auto params = get_test_parameters("testconnection");
  const std::string user = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_USER");
  const std::string password = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_PASSWORD");
  const std::string client_id = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_CLIENT_ID");
  const std::string client_secret =
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_CLIENT_SECRET");
  const std::string redirect_uri =
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_REDIRECT_URI");
  const std::string totp_seed = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_MFA_SEED");

  // Given Authentication is set to OAUTH_AUTHORIZATION_CODE with a valid client id / secret. `oauth_authorization_url`
  // and `oauth_token_request_url` are forwarded from parameters when present (otherwise the driver falls back to the
  // Snowflake-IdP defaults `https://{host}/oauth/authorize` and `https://{host}/oauth/token-request`).
  // `client_store_temporary_credential=true` lets the AC flow short-circuit on subsequent runs by re-using the cached
  // access / refresh token (AC state machine: cache → refresh → interactive).
  std::stringstream ss;
  ss << get_oauth_base_connection_string(params, "OAUTH_AUTHORIZATION_CODE", user);
  ss << "OAUTH_CLIENT_ID=" << client_id << ";";
  ss << "OAUTH_CLIENT_SECRET=" << client_secret << ";";
  ss << "OAUTH_REDIRECT_URI=" << redirect_uri << ";";
  std::string connection_string = ss.str();

  auto env = setup_oauth_environment();
  auto dbc = env.createConnectionHandle();

  oauth_auth::clean_browser_processes();

  // When Trying to Connect (this will spawn the local-loopback HTTP listener and `xdg-open`/`open`/`ShellExecute` the
  // IdP login URL unless a previously cached access token short-circuits the leg)
  SQLRETURN ret = oauth_auth::connect_with_browser_automation(dbc, connection_string, "internalOauthSnowflakeSuccess",
                                                              user, password, totp_seed);
  oauth_auth::clean_browser_processes();

  // Then Login is successful and a simple query can be executed
  REQUIRE_ODBC(ret, dbc);
  verify_oauth_simple_query_execution(dbc);

  SQLRETURN disconnect_ret = SQLDisconnect(dbc.getHandle());
  REQUIRE_ODBC(disconnect_ret, dbc);
}

TEST_CASE("oauth should reuse cached access token without browser interaction", "[oauth_e2e][requires_browser]") {
  REQUIRE_BROWSER("Authorization Code token caching needs the headless Chromium container");

  auto params = get_test_parameters("testconnection");
  const std::string user = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_USER");
  const std::string password = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_PASSWORD");
  const std::string client_id = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_CLIENT_ID");
  const std::string client_secret =
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_CLIENT_SECRET");
  const std::string redirect_uri =
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_REDIRECT_URI");
  const std::string totp_seed = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_MFA_SEED");

  // Given Authentication is set to OAUTH_AUTHORIZATION_CODE with client_store_temporary_credential=true
  //       and a token has been cached from a previous browser authentication
  std::stringstream ss;
  ss << get_oauth_base_connection_string(params, "OAUTH_AUTHORIZATION_CODE", user);
  ss << "OAUTH_CLIENT_ID=" << client_id << ";";
  ss << "OAUTH_CLIENT_SECRET=" << client_secret << ";";
  ss << "OAUTH_REDIRECT_URI=" << redirect_uri << ";";
  ss << "CLIENT_STORE_TEMPORARY_CREDENTIAL=true;";
  std::string connection_string = ss.str();

  auto env = setup_oauth_environment();

  oauth_auth::clean_browser_processes();
  {
    auto first = env.createConnectionHandle();
    SQLRETURN ret = oauth_auth::connect_with_browser_automation(
        first, connection_string, "internalOauthSnowflakeSuccess", user, password, totp_seed);
    oauth_auth::clean_browser_processes();
    REQUIRE_ODBC(ret, first);
    verify_oauth_simple_query_execution(first);

    ret = SQLDisconnect(first.getHandle());
    REQUIRE_ODBC(ret, first);
  }

  // When Trying to Connect without browser interaction
  auto second = env.createConnectionHandle();
  SQLRETURN ret = SQLDriverConnect(second.getHandle(), nullptr, (SQLCHAR*)connection_string.c_str(), SQL_NTS, nullptr,
                                   0, nullptr, SQL_DRIVER_NOPROMPT);
  REQUIRE_ODBC(ret, second);

  // Then Login is successful and a simple query can be executed
  verify_oauth_simple_query_execution(second);

  ret = SQLDisconnect(second.getHandle());
  REQUIRE_ODBC(ret, second);
}

TEST_CASE("oauth should fail authorization code flow with bad client secret", "[oauth_e2e][requires_browser]") {
  SKIP("Disabled: bad-secret tests cause pipeline flakiness by blocking the test account");
  REQUIRE_BROWSER("Authorization Code negative path needs the headless Chromium container");

  auto params = get_test_parameters("testconnection");
  const std::string user = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_USER");
  const std::string password = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_PASSWORD");
  const std::string client_id = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_CLIENT_ID");
  const std::string redirect_uri =
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_REDIRECT_URI");
  const std::string totp_seed = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OAUTH_SNOWFLAKE_MFA_SEED");

  // Given Authentication is set to OAUTH_AUTHORIZATION_CODE with a
  //       valid client id but a deliberately invalid client secret.
  //       The IdP token-exchange step must reject the credentials
  //       and the driver must surface an authentication / login error.
  std::stringstream ss;
  ss << get_oauth_base_connection_string(params, "OAUTH_AUTHORIZATION_CODE", user);
  ss << "OAUTH_CLIENT_ID=" << client_id << ";";
  ss << "OAUTH_CLIENT_SECRET=invalid_client_secret_12345;";  // pragma: allowlist secret
  ss << "OAUTH_REDIRECT_URI=" << redirect_uri << ";";
  ss << "CLIENT_STORE_TEMPORARY_CREDENTIAL=false;";
  std::string connection_string = ss.str();

  auto env = setup_oauth_environment();
  auto dbc = env.createConnectionHandle();

  oauth_auth::clean_browser_processes();

  // When Trying to Connect
  SQLRETURN ret = oauth_auth::connect_with_browser_automation(dbc, connection_string, "internalOauthSnowflakeSuccess",
                                                              user, password, totp_seed);
  oauth_auth::clean_browser_processes();

  // Then Connection fails with an authentication / login error
  REQUIRE(ret == SQL_ERROR);
  auto records = get_diag_rec(dbc);
  REQUIRE(records.size() >= 1);
  // BD#83: the new driver maps an IdP token-exchange rejection to SQLSTATE 28000
  // (invalid authorization specification); the legacy driver surfaces it as the
  // generic HY000.
  OLD_DRIVER_ONLY("BD#83") { CHECK(records[0].sqlState == "HY000"); }
  NEW_DRIVER_ONLY("BD#83") { CHECK(records[0].sqlState == "28000"); }
}

// =============================================================================
// OAuth Client Credentials (CC) flow
//
// A non-interactive, machine-to-machine flow where an external IdP mints the
// token from a client id / secret. Snowflake's GS does not mint CC tokens, so
// OAUTH_TOKEN_REQUEST_URL is required up-front.
// =============================================================================

TEST_CASE("oauth should authenticate using client credentials flow", "[oauth_e2e][requires_browser]") {
  REQUIRE_BROWSER("OAuth E2E needs the headless browser container (preprod IdP connectivity)");

  auto params = get_test_parameters("testconnection");
  const std::string token_url = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_TOKEN_URL");
  const std::string client_id = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_EXTERNAL_CLIENT_ID");
  const std::string client_secret =
      get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_EXTERNAL_CLIENT_SECRET");

  // Given Authentication is set to OAUTH_CLIENT_CREDENTIALS with a valid client id / secret and an external IdP token
  // URL. Snowflake's GS does not mint CC tokens, so `oauth_token_request_url` is required up-front.
  std::stringstream ss;
  ss << get_oauth_base_connection_string(params, "OAUTH_CLIENT_CREDENTIALS", client_id);
  ss << "OAUTH_CLIENT_ID=" << client_id << ";";
  ss << "OAUTH_CLIENT_SECRET=" << client_secret << ";";
  ss << "OAUTH_TOKEN_REQUEST_URL=" << token_url << ";";
  std::string connection_string = ss.str();

  auto env = setup_oauth_environment();
  auto dbc = env.createConnectionHandle();

  // When Trying to Connect
  SQLRETURN ret = SQLDriverConnect(dbc.getHandle(), nullptr, (SQLCHAR*)connection_string.c_str(), SQL_NTS, nullptr, 0,
                                   nullptr, SQL_DRIVER_NOPROMPT);
  REQUIRE_ODBC(ret, dbc);

  // Then Login is successful and a simple query can be executed
  verify_oauth_simple_query_execution(dbc);

  SQLRETURN disconnect_ret = SQLDisconnect(dbc.getHandle());
  REQUIRE_ODBC(disconnect_ret, dbc);
}

TEST_CASE("oauth should fail client credentials flow with bad client secret", "[oauth_e2e][requires_browser]") {
  SKIP("Disabled: bad-secret tests cause pipeline flakiness by blocking the test account");
  REQUIRE_BROWSER("OAuth E2E needs the headless browser container (preprod IdP connectivity)");

  auto params = get_test_parameters("testconnection");
  const std::string token_url = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_TOKEN_URL");
  const std::string client_id = get_param_required<std::string>(params, "SNOWFLAKE_TEST_OKTA_OAUTH_EXTERNAL_CLIENT_ID");

  // Given Authentication is set to OAUTH_CLIENT_CREDENTIALS with a valid client id, an invalid client secret and a
  // valid token_request_url
  std::stringstream ss;
  ss << get_oauth_base_connection_string(params, "OAUTH_CLIENT_CREDENTIALS", client_id);
  ss << "OAUTH_CLIENT_ID=" << client_id << ";";
  ss << "OAUTH_CLIENT_SECRET=invalid_client_secret_12345;";  // pragma: allowlist secret
  ss << "OAUTH_TOKEN_REQUEST_URL=" << token_url << ";";
  std::string connection_string = ss.str();

  ensure_driver_installed();

  // When Trying to Connect
  auto records = require_connection_failed(connection_string);

  // Then Connection fails with an authentication / login error
  REQUIRE(records.size() >= 1);
  // BD#83: the new driver maps an IdP token-exchange rejection to SQLSTATE 28000
  // (invalid authorization specification); the legacy driver surfaces it as the
  // generic HY000.
  OLD_DRIVER_ONLY("BD#83") { CHECK(records[0].sqlState == "HY000"); }
  NEW_DRIVER_ONLY("BD#83") { CHECK(records[0].sqlState == "28000"); }
}

File: sf_core/tests/e2e/authentication/oauth.rs
//! End-to-end OAuth tests against a real Snowflake account / IdP.
//!
//! These tests are feature-gated behind `auth_oauth_e2e` (the entire
//! module is `#[cfg(feature = "auth_oauth_e2e")] mod oauth;` in
//! `tests/e2e/authentication/mod.rs`) and `#[ignore]`d so they do not
//! run in CI by default — most need real OAuth client credentials and
//! at least one needs a live browser leg. They mirror the `vpn_*`
//! gating pattern from `native_okta.rs` and the feature-gating pattern
//! from `user_password_mfa.rs`, so a developer can opt in selectively.
//!
//! ## How to run
//!
//! ```bash
//! # 1. Make sure your parameters.json has the OAuth fields (see below).
//! export PARAMETER_PATH=/path/to/parameters.json
//!
//! # 2. Run all OAuth E2E tests (the `oauth_` prefix targets just this
//! #    module):
//! cargo test --package sf_core --features auth_oauth_e2e --release \
//!     --test e2e_tests -- --ignored oauth_
//!
//! # 3. Or pick a single test:
//! cargo test --package sf_core --features auth_oauth_e2e --release \
//!     --test e2e_tests -- --ignored \
//!     oauth_should_authenticate_with_pre_acquired_access_token
//! ```
//!
//! ## Required `parameters.json` fields (`testconnection` block)
//!
//! All `SNOWFLAKE_TEST_OAUTH_*` keys are optional; tests that need a
//! particular field either `expect()` it (hard fail when the parameter
//! is genuinely required) or `eprintln!` and early-return when the
//! parameter is only sometimes available (e.g. the CC-flow token URL,
//! which requires an external IdP). The cross-driver names follow
//! `analysis_feature_oauth.md` §9:
//!
//! | parameters.json key                           | drives                                       |
//! |-----------------------------------------------|----------------------------------------------|
//! | `SNOWFLAKE_TEST_OAUTH_CLIENT_ID`              | AC + CC: OAuth client id (required for both) |
//! | `SNOWFLAKE_TEST_OAUTH_CLIENT_SECRET`          | AC + CC: OAuth client secret                 |
//! | `SNOWFLAKE_TEST_OAUTH_AUTHORIZATION_URL`      | AC: IdP authorization endpoint (optional;    |
//! |                                               | defaults to `https://{host}/oauth/authorize`)|
//! | `SNOWFLAKE_TEST_OAUTH_TOKEN_REQUEST_URL`      | AC (optional) + CC (required): IdP token URL |
//! | `SNOWFLAKE_TEST_OAUTH_REDIRECT_URI`           | AC: loopback redirect URI (optional)         |
//! | `SNOWFLAKE_TEST_OAUTH_SCOPE`                  | AC + CC: requested OAuth scope (optional)    |
//! | `SNOWFLAKE_TEST_OAUTH_ACCESS_TOKEN`           | legacy `AUTHENTICATOR=OAUTH` flow            |

use crate::common::config::Parameters;
use crate::common::snowflake_test_client::SnowflakeTestClient;
use sf_core::token_cache::{KeyringTokenCache, TokenCache, TokenType};

// =============================================================================
// Legacy `AUTHENTICATOR=OAUTH` (analysis §6 / §10.1)
// =============================================================================

#[test]
#[ignore = "OAuth E2E: requires SNOWFLAKE_TEST_OAUTH_ACCESS_TOKEN (run with --features auth_oauth_e2e)"]
fn oauth_should_authenticate_with_pre_acquired_access_token() {
    // Given Authentication is set to legacy OAUTH and a pre-acquired
    //       OAuth access token is supplied via `token=`
    let client = SnowflakeTestClient::with_default_params();
    let access_token = client
        .parameters
        .oauth_access_token
        .clone()
        .expect("SNOWFLAKE_TEST_OAUTH_ACCESS_TOKEN must be set for legacy OAUTH E2E test");

    set_legacy_oauth(&client);
    set_oauth_token(&client, &access_token);

    // When Trying to Connect
    let result = client.connect();

    // Then Login is successful and a simple query can be executed
    client.verify_simple_query(result);
}

#[test]
#[ignore = "OAuth E2E: legacy OAUTH negative path (run with --features auth_oauth_e2e)"]
fn oauth_should_fail_legacy_authentication_with_invalid_token() {
    // Given Authentication is set to legacy OAUTH and an invalid
    //       OAuth access token is supplied
    let client = SnowflakeTestClient::with_default_params();
    set_legacy_oauth(&client);
    set_oauth_token(&client, "invalid_oauth_token_12345");

    // When Trying to Connect
    let result = client.connect();

    // Then Connection fails with an authentication / login error
    client.assert_login_error(result);
}

// =============================================================================
// OAuth Authorization Code (AC) flow (analysis §3 / §10.2)
// =============================================================================

#[test]
#[ignore = "OAuth E2E: AC flow needs a real browser leg unless an OAuth access token is already cached in the keyring (run with --features auth_oauth_e2e)"]
fn oauth_should_authenticate_using_authorization_code_flow() {
    // Given Authentication is set to OAUTH_AUTHORIZATION_CODE with a
    //       valid client id / secret. `oauth_authorization_url` and
    //       `oauth_token_request_url` are forwarded from parameters
    //       when present (otherwise the driver falls back to the
    //       Snowflake-IdP defaults `https://{host}/oauth/authorize`
    //       and `https://{host}/oauth/token-request`).
    //       `client_store_temporary_credential=true` lets the AC flow
    //       short-circuit on subsequent runs by re-using the cached
    //       access / refresh token (analysis §3.2 / §7).
    let client = SnowflakeTestClient::with_default_params();
    let client_id = client
        .parameters
        .oauth_client_id
        .clone()
        .expect("SNOWFLAKE_TEST_OAUTH_CLIENT_ID must be set for OAuth AC E2E test");
    let client_secret = client
        .parameters
        .oauth_client_secret
        .clone()
        .expect("SNOWFLAKE_TEST_OAUTH_CLIENT_SECRET must be set for OAuth AC E2E test");

    set_authorization_code(&client);
    client.set_connection_option("oauth_client_id", &client_id);
    client.set_connection_option("oauth_client_secret", &client_secret);
    set_optional_oauth_endpoints(&client);
    client.set_connection_option_bool("client_store_temporary_credential", true);

    // When Trying to Connect (this will spawn the local-loopback HTTP
    // listener and `xdg-open`/`open`/`ShellExecute` the IdP login URL
    // unless a previously cached access token short-circuits the leg)
    let result = client.connect();

    // Then Login is successful and a simple query can be executed
    client.verify_simple_query(result);
}

#[test]
#[ignore = "OAuth E2E: AC short-circuit via cached access token (run with --features auth_oauth_e2e)"]
fn oauth_should_short_circuit_authorization_code_flow_with_cached_access_token() {
    // Given Authentication is set to OAUTH_AUTHORIZATION_CODE and a
    //       valid OAuth access token is pre-seeded in the OS keyring
    //       under the (host, user, OAUTH_ACCESS_TOKEN) cache key. The
    //       host is derived from `oauth_token_request_url` — falling
    //       back to the Snowflake server URL — exactly like
    //       `host_from_token_url` in production code (analysis §7.3).
    let client = SnowflakeTestClient::with_default_params();
    let client_id = client
        .parameters
        .oauth_client_id
        .clone()
        .expect("SNOWFLAKE_TEST_OAUTH_CLIENT_ID must be set for OAuth AC short-circuit test");
    let client_secret =
        client.parameters.oauth_client_secret.clone().expect(
            "SNOWFLAKE_TEST_OAUTH_CLIENT_SECRET must be set for OAuth AC short-circuit test",
        );
    let access_token =
        client.parameters.oauth_access_token.clone().expect(
            "SNOWFLAKE_TEST_OAUTH_ACCESS_TOKEN must be set for OAuth AC short-circuit test",
        );
    let user = client
        .parameters
        .user
        .clone()
        .expect("SNOWFLAKE_TEST_USER must be set for OAuth AC short-circuit test");
    let cache_host = derive_cache_host(&client.parameters)
        .expect("expected to derive a cache host from oauth_token_request_url or server URL");

    // Pre-seed the OAuth access token so the AC flow skips the
    // browser leg entirely (mirroring the wiremock-driven
    // short-circuit test in `tests/integration/authentication/oauth.rs`
    // and the keyring-based pattern from
    // `tests/integration/authentication/user_password_mfa_token_cache.rs`).
    let cache = KeyringTokenCache::new().expect("keyring token cache should be available");
    cache
        .add_token(
            &cache_host,
            &user,
            TokenType::OAuthAccessToken,
            &access_token,
        )
        .expect("seed OAuth access token");

    set_authorization_code(&client);
    client.set_connection_option("oauth_client_id", &client_id);
    client.set_connection_option("oauth_client_secret", &client_secret);
    set_optional_oauth_endpoints(&client);
    client.set_connection_option_bool("client_store_temporary_credential", true);

    // When Trying to Connect — should NOT spawn a browser; the
    // pre-seeded access token must satisfy the AC short-circuit.
    let result = client.connect();

    // Then Login is successful and a simple query can be executed
    client.verify_simple_query(result);

    // Cleanup: remove the cache entries we seeded so the keyring
    // doesn't accumulate test artefacts even if the test panicked
    // earlier (best-effort; we ignore errors).
    let _ = cache.remove_token(&cache_host, &user, TokenType::OAuthAccessToken);
    let _ = cache.remove_token(&cache_host, &user, TokenType::OAuthRefreshToken);
}

// =============================================================================
// OAuth Client Credentials (CC) flow (analysis §4 / §10.3)
// =============================================================================

#[test]
#[ignore = "OAuth E2E: CC flow requires an external IdP and a configured oauth_token_request_url (run with --features auth_oauth_e2e)"]
fn oauth_should_authenticate_using_client_credentials_flow() {
    // Given Authentication is set to OAUTH_CLIENT_CREDENTIALS with a
    //       valid client id / secret and an external IdP token URL.
    //       Snowflake's GS does not mint CC tokens (analysis §4), so
    //       `oauth_token_request_url` is required up-front.
    let client = SnowflakeTestClient::with_default_params();
    let client_id = client
        .parameters
        .oauth_client_id
        .clone()
        .expect("SNOWFLAKE_TEST_OAUTH_CLIENT_ID must be set for OAuth CC E2E test");
    let client_secret = client
        .parameters
        .oauth_client_secret
        .clone()
        .expect("SNOWFLAKE_TEST_OAUTH_CLIENT_SECRET must be set for OAuth CC E2E test");
    let Some(token_url) = client.parameters.oauth_token_request_url.clone() else {
        eprintln!(
            "Skipping oauth_should_authenticate_using_client_credentials_flow: \
             SNOWFLAKE_TEST_OAUTH_TOKEN_REQUEST_URL is not configured (CC requires an external IdP)"
        );
        return;
    };

    set_client_credentials(&client);
    client.set_connection_option("oauth_client_id", &client_id);
    client.set_connection_option("oauth_client_secret", &client_secret);
    client.set_connection_option("oauth_token_request_url", &token_url);
    if let Some(scope) = client.parameters.oauth_scope.clone() {
        client.set_connection_option("oauth_scope", &scope);
    }

    // When Trying to Connect
    let result = client.connect();

    // Then Login is successful and a simple query can be executed
    client.verify_simple_query(result);
}

// =============================================================================
// AC flow — bad client secret (analysis §13: IdP-error surface)
// =============================================================================

#[test]
#[ignore = "OAuth E2E: AC negative path with a bad client secret (run with --features auth_oauth_e2e)"]
fn oauth_should_fail_authorization_code_flow_with_bad_client_secret() {
    // Given Authentication is set to OAUTH_AUTHORIZATION_CODE with a
    //       valid client id but a deliberately invalid client secret.
    //       The IdP token-exchange step must reject the credentials
    //       and the driver must surface an authentication / login
    //       error.
    let client = SnowflakeTestClient::with_default_params();
    let client_id = client
        .parameters
        .oauth_client_id
        .clone()
        .expect("SNOWFLAKE_TEST_OAUTH_CLIENT_ID must be set for OAuth AC negative E2E test");

    set_authorization_code(&client);
    client.set_connection_option("oauth_client_id", &client_id);
    client.set_connection_option("oauth_client_secret", "invalid_client_secret_12345"); // pragma: allowlist secret
    set_optional_oauth_endpoints(&client);
    client.set_connection_option_bool("client_store_temporary_credential", false);

    // When Trying to Connect
    let result = client.connect();

    // Then Connection fails with an authentication / login error
    client.assert_login_error(result);
}

// =============================================================================
// Helpers
// =============================================================================

fn set_legacy_oauth(client: &SnowflakeTestClient) {
    client.set_connection_option("authenticator", "OAUTH");
}

fn set_authorization_code(client: &SnowflakeTestClient) {
    client.set_connection_option("authenticator", "OAUTH_AUTHORIZATION_CODE");
}

fn set_client_credentials(client: &SnowflakeTestClient) {
    client.set_connection_option("authenticator", "OAUTH_CLIENT_CREDENTIALS");
}

fn set_oauth_token(client: &SnowflakeTestClient, token: &str) {
    client.set_connection_option("token", token);
}

/// Forward optional AC endpoints (`oauth_authorization_url`,
/// `oauth_token_request_url`, `oauth_redirect_uri`, `oauth_scope`) from
/// `parameters.json`. When the IdP is Snowflake itself the driver
/// derives the authorization / token URLs from the host, so leaving
/// these unset is valid.
fn set_optional_oauth_endpoints(client: &SnowflakeTestClient) {
    if let Some(url) = client.parameters.oauth_authorization_url.clone() {
        client.set_connection_option("oauth_authorization_url", &url);
    }
    if let Some(url) = client.parameters.oauth_token_request_url.clone() {
        client.set_connection_option("oauth_token_request_url", &url);
    }
    if let Some(uri) = client.parameters.oauth_redirect_uri.clone() {
        client.set_connection_option("oauth_redirect_uri", &uri);
    }
    if let Some(scope) = client.parameters.oauth_scope.clone() {
        client.set_connection_option("oauth_scope", &scope);
    }
}

/// Mirror of `sf_core::rest::snowflake::oauth::host_from_token_url`:
/// the OAuth token cache keys off the IdP token endpoint host when
/// available, otherwise the Snowflake server URL host (Python-style
/// `urlparse(token_request_url).hostname`, analysis §7.3). Kept private
/// to this test file so the production helper does not need to be
/// promoted to `pub`.
fn derive_cache_host(parameters: &Parameters) -> Option<String> {
    fn host_of(raw: &str) -> Option<String> {
        url::Url::parse(raw)
            .ok()
            .and_then(|u| u.host_str().map(str::to_string))
    }
    if let Some(token_url) = parameters.oauth_token_request_url.as_deref()
        && let Some(host) = host_of(token_url)
    {
        return Some(host);
    }
    if let Some(server_url) = parameters.get_server_url()
        && let Some(host) = host_of(&server_url)
    {
        return Some(host);
    }
    parameters.host.clone()
}
